fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/DefaultCustomIngredients.java

```diff
@@ -0,0 +1,134 @@
+import net.fabricmc.fabric.impl.ingredient.builtin.NbtIngredient;
+import net.fabricmc.fabric.impl.ingredient.builtin.OrIngredient;
+
+/**
+ * Factory methods for the custom ingredients directly provided by fabric.
+ */
+public final class DefaultCustomIngredients {
+	/**
+	 * Create an ingredient that matches what its sub-ingredients all match.
+	 *
+	 * <p>The JSON format is as follows:
+	 * <pre>{@code
+	 * {
+	 *     "fabric:type": "fabric:and",
+	 *     "ingredients": [
+	 *         // sub-ingredient 1,
+	 *         // sub-ingredient 2,
+	 *         // etc...
+	 *     ]
+	 * }
+	 * }</pre>
+	 */
+	public static Ingredient and(Ingredient... ingredients) {
+		for (Ingredient ing : ingredients) Objects.requireNonNull(ing);
+
+		return new AndIngredient(ingredients).toVanilla();
+	}
+
+	/**
+	 * Create an ingredient that matches what any of its sub-ingredients matches.
+	 *
+	 * <p>The JSON format is as follows:
+	 * <pre>{@code
+	 * {
+	 *     "fabric:type": "fabric:or",
+	 *     "ingredients": [
+	 *         // sub-ingredient 1,
+	 *         // sub-ingredient 2,
+	 *         // etc...
+	 *     ]
+	 * }
+	 * }</pre>
+	 */
+	public static Ingredient or(Ingredient... ingredients) {
+		for (Ingredient ing : ingredients) Objects.requireNonNull(ing);
+
+		return new OrIngredient(ingredients).toVanilla();
+	}
+
+	/**
+	 * Create an ingredient that matches if its base ingredient matches, and its subtracted ingredient does NOT match.
+	 *
+	 * <p>The JSON format is as follows:
+	 * <pre>{@code
+	 * {
+	 *     "fabric:type": "fabric:difference",
+	 *     "base": // base ingredient,
+	 *     "subtracted": // subtracted ingredient
+	 * }
+	 * }</pre>
+	 */
+	public static Ingredient difference(Ingredient base, Ingredient subtracted) {
+		Objects.requireNonNull(base);
+		Objects.requireNonNull(subtracted);
+
+		return new DifferenceIngredient(base, subtracted).toVanilla();
+	}
+
+	/**
+	 * Create an ingredient that wraps another ingredient to also check for stack NBT.
+	 * This check can either be strict (the exact NBT must match) or non-strict (the ingredient NBT must be a subset of the stack NBT).
+	 *
+	 * <p>Note: refer to {@link NbtHelper#matches} for the non-strict matching.
+	 *
+	 * <p>The JSON format is as follows:
+	 * <pre>{@code
+	 * {
+	 *    "fabric:type": "fabric:nbt",
+	 *    "base": // base ingredient,
+	 *    "nbt": // NBT tag to match, either in JSON directly or a string representation (default: null),
+	 *    "strict": // whether to use strict matching (default: false)
+	 * }
+	 * }</pre>
+	 */
+	public static Ingredient nbt(Ingredient base, @Nullable NbtCompound nbt, boolean strict) {
+		Objects.requireNonNull(base);
+
+		return new NbtIngredient(base, nbt, strict).toVanilla();
+	}
+
+	/**
+	 * Create an ingredient that matches the passed template stack, NBT included.
+	 * This check can either be strict (the exact NBT must match) or non-strict (the template NBT must be a subset of the stack NBT).
```

Comment:
Document that this crashes with NBT-less stack and non-strict mode. Or rework the specs on this.